Source/DafnyLanguageServer/Workspace/Document.cs

```diff
@@ -0,0 +1,174 @@
+  ///
+  /// When verification starts, no new instances of Compilation will be created for this version.
+  /// There can be different verification threads that update the state of this object.
+  /// </summary>
+  public class Document {
+    public DocumentTextBuffer TextDocumentItem { get; }
+    public DocumentUri Uri => TextDocumentItem.Uri;
+    public int Version => TextDocumentItem.Version!.Value;
+
+    public Document(DocumentTextBuffer textDocumentItem) {
+      TextDocumentItem = textDocumentItem;
+    }
+
+    public virtual IEnumerable<Diagnostic> Diagnostics => Enumerable.Empty<Diagnostic>();
+
+    public IdeState InitialIdeState() {
+      return ToIdeState(new IdeState(TextDocumentItem, Array.Empty<Diagnostic>(),
+        SymbolTable.Empty(TextDocumentItem), new Dictionary<ImplementationId, ImplementationView>(),
+        false, Array.Empty<Diagnostic>(),
+        new DocumentVerificationTree(TextDocumentItem)));
+    }
+
+    /// <summary>
+    /// Collects information to present to the IDE
+    /// </summary>
+    public virtual IdeState ToIdeState(IdeState previousState) {
+      return previousState with {
+        TextDocumentItem = TextDocumentItem,
+        ImplementationsWereUpdated = false,
+      };
+    }
+  }
+
+  public class DocumentAfterParsing : Document {
+    private readonly IReadOnlyList<Diagnostic> parseDiagnostics;
+
+    public DocumentAfterParsing(DocumentTextBuffer textDocumentItem,
+      Dafny.Program program,
+      IReadOnlyList<Diagnostic> parseDiagnostics) : base(textDocumentItem) {
+      this.parseDiagnostics = parseDiagnostics;
+      Program = program;
+    }
+
+    public override IEnumerable<Diagnostic> Diagnostics => parseDiagnostics;
+
+    public Dafny.Program Program { get; }
+
+    public override IdeState ToIdeState(IdeState previousState) {
+      return previousState with {
+        TextDocumentItem = TextDocumentItem,
+        ResolutionDiagnostics = parseDiagnostics,
+        ImplementationsWereUpdated = false,
+      };
+    }
+  }
+
+  public class DocumentAfterResolution : DocumentAfterParsing {
+    public DocumentAfterResolution(DocumentTextBuffer textDocumentItem,
+      Dafny.Program program,
+      IReadOnlyList<Diagnostic> parseAndResolutionDiagnostics,
+      SymbolTable symbolTable,
+      IReadOnlyList<Diagnostic> ghostDiagnostics) :
+      base(textDocumentItem, program, ArraySegment<Diagnostic>.Empty) {
+      ParseAndResolutionDiagnostics = parseAndResolutionDiagnostics;
+      SymbolTable = symbolTable;
+      GhostDiagnostics = ghostDiagnostics;
+    }
+
+    public IReadOnlyList<Diagnostic> ParseAndResolutionDiagnostics { get; }
+    public SymbolTable SymbolTable { get; }
+    public IReadOnlyList<Diagnostic> GhostDiagnostics { get; }
+
+    public override IEnumerable<Diagnostic> Diagnostics => ParseAndResolutionDiagnostics;
+
+    public override IdeState ToIdeState(IdeState previousState) {
```

Comment:
Excellent, that makes much more sense now that we use a document to update an existing IdeState